File: github_code/avl-tree/java/9.java
package analysis.in.java.chapter4;

import analysis.in.java.util.UnderflowException;

public class AvlTree <AnyType extends Comparable<? super AnyType>>{

	public static enum Traveral{
		PRE_ORDER,IN_ORDER,POST_ORDER
	}
	private static class AvlNode<AnyType>{
		AnyType element;
		AvlNode<AnyType> left;
		AvlNode<AnyType> right;
		int height;
		
		public AvlNode(AnyType element){
			this(element,null,null);
		}
		
		public AvlNode(AnyType element,AvlNode<AnyType> left,
				AvlNode<AnyType> right){
			this.element=element;
			this.left=left;
			this.right=right;
		}
	}
	
	private AvlNode<AnyType> root;
	
	public AvlTree(){
		root=null;
	}
	
	public void makeEmpty(){
		root=null;
	}
	
	public boolean isEmpty(){
		return null==root;
	}
	
	public void printTree(Traveral type){
		printTree(root,type);
	}
	/**
	 * Internal method to print a subtree in sorted order
	 * @param t the node that roots the subtree
	 */
	private void printTree(AvlNode<AnyType> t,Traveral type){
		if(null!=t){
			switch(type){
			case PRE_ORDER:
				printName(t.element,t.height);
				printTree(t.left,type);
				printTree(t.right,type);
				break;
			case IN_ORDER:
				printTree(t.left,type);
				printName(t.element,t.height);
				printTree(t.right,type);
				break;
			case POST_ORDER:
				printTree(t.left,type);
				printTree(t.right,type);
				printName(t.element,t.height);
				break;
			}
		}
	}
	private void printName(AnyType element,int depth){
		System.out.println(element);
	}
	
	public void insert(AnyType element){
		root=insert(element,root);
	}
	/**
	 * Internal method to insert into a subtree.
	 * @param element the item to insert.
	 * @param t the node that roots the subtree.
	 * @return the new root of the subtree.
	 */
	private AvlNode<AnyType> insert(AnyType element,AvlNode<AnyType> t){
		if(null==t){
			return new AvlNode<AnyType>(element);
		}
		int compareResult=element.compareTo(t.element);
		if(compareResult<0){
			t.left=insert(element,t.left);
			//  maybe can call rotate instead of those code
//			if(height(t.left)-height(t.right) == 2){
//				if(element.compareTo(t.left.element)<0){
//					t=rotateWithLeftChild(t);
//				}else {
//					t=doubleWithLeftChild(t);
//				}
//			}
		}else if(compareResult>0){
			t.right=insert(element,t.right);
//			if(height(t.right)-height(t.left) == 2){
//				if(element.compareTo(t.right.element)>0){
//					t=rotateWithRightChild(t);
//				}else {
//					t=doubleWithRightChild(t);
//				}
//			}
		}else{
			//Duplicate ;do nothing
		}
		t=rotate(t);
		updateHeight(t);
		return t;
	}
	
	
	public void remove(AnyType element){
		root=remove(element,root);
	}
	
	/**
	 * 删除一个节点的主要步骤如下所示：
	 * 1.如果是null节点，则直接返回null
	 * 2.如果删除的值<当前节点，则转入left节点进行递归删除
	 * 3.如果删除的值>当前节点，则转入right节点进行递归删除
	 * 4.如果删除的值为当前节点，如果当前节点只有一个子树，则直接返回该子树
	 * 5.如果删除的值为当前节点，且当前节点有两个子树，则将当前值更改为右子树中最小的节点值，并递归在右子树中删除该节点值
	 * 6.重新修正该处理节点的height值
	 * 7.对处理节点进行重新翻转处理，以修正在删除过程中可能出现的树不平衡情况
	 * @param element
	 * @param t
	 * @return
	 */
	private  AvlNode<AnyType> remove(AnyType element, AvlNode<AnyType> t) {
        if(t == null)
            return null;
        int cmp = element.compareTo(t.element);
        if(cmp < 0) {
            t.left = remove(element, t.left);
        } else if(cmp > 0) {
            t.right = remove(element, t.right);
        } else if(t.left != null && t.right != null) {
            t.element = findMin(t.right).element;
            t.right = remove(t.element, t.right);
        } else {
            t = t.left == null ? t.right : t.left;
        }
        t = rotate(t);
        updateHeight(t);
        return t;
    }
	public boolean contains(AnyType element){
		return contains(element,root);
	}
	/**
	 * Internal method to search an item in a subtree
	 * @param element the item to search for.
	 * @param t the node that roots the subtree. 
	 * @return whether the subtree has a node that containing the search element.
	 */
	private boolean contains(AnyType element,AvlNode<AnyType> t){
		if(null==t){
			return false;
		}
		int compareResult=element.compareTo(t.element);
		if(compareResult<0){
			return contains(element,t.left);
		}else if(compareResult>0){
			return contains(element,t.right);
		}
		//Match
		return true;
	}
	public AnyType findMin(){
		if(null==root){
			throw new UnderflowException();
		}
		return findMin(root).element;
	}
	/**
	 * Internal method to find the smallest item in a subtree.
	 * @param t the node that roots the subtree.
	 * @return node that containing the smallest item.
	 */
	private AvlNode<AnyType> findMin(AvlNode<AnyType> t){
		if(null!=t){
			while(t.left!=null){
				t=t.left;
			}
		}
		return t;
	}
	public AnyType findMax(){
		if(null==root){
			throw new UnderflowException();
		}
		return findMax(root).element;
	}
	/**
	 * Internal method to find the largest item in a subtree.
	 * @param t the node that roots the subtree.
	 * @return node that containing the largest item.
	 */
	private AvlNode<AnyType> findMax(AvlNode<AnyType> t){
		if(null!=t){
			while(t.right!=null){
				t=t.right;
			}
		}
		return t;
	}
	/**
	 * Internal method to get the height of a node
	 * @param t
	 * @return the height of node t,or -1,if null.
	 */
	private int height(AvlNode<AnyType> t){
		return null==t?-1:t.height;
	}
	
	/**
	 * Internal method to update the height of a node.
	 * you should update t's left and right child first to update t.
	 * @param t the node that roots the subtree.
	 */
	private void updateHeight(AvlNode<AnyType> t){
		t.height=Math.max(height(t.left),height(t.right))+1;
	}


	
	private  AvlNode<AnyType> rotate( AvlNode<AnyType> t) {
        if(t == null)
            return null;
        if(height(t.left) - height(t.right) == 2) {
            if(height(t.left.left) >= height(t.left.right))
                return rotateWithLeftChild(t);
            else
                return doubleWithLeftChild(t);
        } else if(height(t.right) - height(t.left) == 2) {
            if(height(t.right.right) >= height(t.right.left))
                return rotateWithRightChild(t);
            else
                return doubleWithRightChild(t);
        }
        return t;
    }
	/**
	 * Rotate binary tree node with left child. 
	 * For AVL trees, this is a single rotation for case 1
	 * Update heights,then return new root.
	 */
	private AvlNode<AnyType> rotateWithLeftChild(AvlNode<AnyType> k2){
		AvlNode<AnyType> k1=k2.left;
		k2.left=k1.right;
		k1.right=k2;
		updateHeight(k2);
		updateHeight(k1);
		return k1;
	}
	/**
	 * Rotate binary tree node with right child. 
	 * For AVL trees, this is a single rotation for case 3
	 * Update heights,then return new root.
	 */
	private AvlNode<AnyType> rotateWithRightChild(AvlNode<AnyType> k2){
		AvlNode<AnyType> k1=k2.right;
		k2.right=k1.left;
		k1.left=k2;
		updateHeight(k2);
		updateHeight(k1);
		return k1;
	}
	/**
	 * Double rotate binary tree node: first left child 
	 * with its right child;then node k3 with new left child. 
	 * For AVL trees, this is a double rotation for case 2.
	 * Update heights,then return new root.
	 */
	private AvlNode<AnyType> doubleWithLeftChild(AvlNode<AnyType> k3){
		k3.left=rotateWithRightChild(k3.left);
		return rotateWithLeftChild(k3);
	}
	
	/**
	 * Double rotate binary tree node: first right child 
	 * with its left child; then node k3 with new right child.
	 * For AVL trees, this is a double rotation for case 4.
	 * Update heights, then return new root
	 */
	private  AvlNode<AnyType> doubleWithRightChild(AvlNode<AnyType> k3){
		k3.right=rotateWithLeftChild(k3.right);
		return rotateWithRightChild(k3);
	}
	




}

File: github_code/avl-tree/cpp/9.c
/*
A program, to Create, Edit and Update/Modify an AVL-Tree.
-Balancing BST, Normal properties, except, the Left Sub-Tree can have atmost length 1
greater than the Right sub-tree, ad this property follows recursiely, if void, a rotation takes place in 
the sub-tree.

Rotation_Explanations: (http://pages.cs.wisc.edu/~paton/readings/liblitVersion/AVL-Tree-Rotations.pdf)
|
Practice Question number:Q6.
Ankit Vadehra
ankitvad@gmail.com
*/

#include <iostream>
using namespace std;

/*
Tree-Node-Decleration
*/
template <class T>  //...Data-Type
struct AVLNode
{
T data;
AVLNode *left; 
AVLNode *right;
int ht;  //...Height of the Sub-Tree, at any AVLNode.
};

//AVLNode<int> *root;

template <class T>  //...Data-Type
class AVLTree
{
public:
	AVLNode<T> *root;
	AVLTree()
	{
		root=NULL;
	};

	int height(AVLNode<T> *);
	int height_difference(AVLNode<T> *);  //...To calclate the Balancing_Factor

	/*
	The 4 Rotation Coditions:
	1)right_right_rotation
	2)left_left_rotation
	3)left_right_rotation
	4)right_left_rotation
	*/
	AVLNode<T> *right_right_rotation(AVLNode<T> *);
	AVLNode<T> *left_left_rotation(AVLNode<T> *);
	AVLNode<T> *left_right_rotation(AVLNode<T> *);
	AVLNode<T> *right_left_rotation(AVLNode<T> *);
	//AVLNode<T> *balance_tree(AVLNode<T> *);  //...For after deleting a AVLNode.
	AVLNode<T> *insert(AVLNode<T> *, const T &key);
	AVLNode<T> *lookup(AVLNode<T> *, const T &key);  //...Using Inorder Traversal/
	void *inorder(AVLNode<T> *);
	//void remove(const T &key);
	AVLNode<T> *remove(AVLNode<T> *, const T &key);
	
	

};

/*
...Calculating the Height
*/
template <class T>  //...Data-Type
int AVLTree<T>::height(AVLNode<T> *temp)
{
	int left_height,right_height;
	  //...Checking for Null Condition, on existing AVLNode.
	if(temp==0)
		{
			return(0);
		}
		//...For Left Node.
	if(temp->left==NULL)
		{left_height=0;}
	else
		left_height=1+(temp->left->ht);
	  //...For Right AVLNode.
	if(temp->right==NULL)
		{right_height=0;}
	else
		right_height=1+(temp->right->ht);
	  //...Maximum Height is Choosen, as Node Height
	if(left_height>right_height)
		return(left_height);
	else
		return(right_height);
}
template <class T>
int AVLTree<T>::height_difference(AVLNode<T> *temp)
{
	  //...Calculation of the Balancing_Factor ie.(LH-RH)
	int left_height=height(temp->left);
	int right_height=height(temp->right);
	int balancing_factor=left_height - right_height;
	return balancing_factor;
}

  //...Right-Right-Rotation(1)
template <class T>
AVLNode<T> *AVLTree<T>::right_right_rotation(AVLNode<T> *parent_AVLNode)
{
	AVLNode<T> *temp;
	temp=parent_AVLNode->right;
	parent_AVLNode->right=temp->left;
	temp->left=parent_AVLNode;
	parent_AVLNode->ht=height(parent_AVLNode);
	temp->ht=height(temp);
	return(temp);
}

  //...Left-Left-Rotation(2)
template <class T>
AVLNode<T> *AVLTree<T>::left_left_rotation(AVLNode<T> *parent_AVLNode)
{
	AVLNode<T> *temp;
	temp=parent_AVLNode->left;
	parent_AVLNode->left=temp->right;
	temp->right=parent_AVLNode;
	parent_AVLNode->ht=height(parent_AVLNode);
	temp->ht=height(temp);
	return(temp);
}

  //...Left-Right-Rotation(3)
template <class T>
AVLNode<T> *AVLTree<T>::left_right_rotation(AVLNode<T> *parent_AVLNode)
{
	AVLNode<T> *temp;
	temp=parent_AVLNode->left;
	parent_AVLNode->left=right_right_rotation(temp);
	return left_left_rotation(parent_AVLNode);
}

//...Right-Left-Rotation(4)
template <class T>
AVLNode<T> *AVLTree<T>::right_left_rotation(AVLNode<T> *parent_AVLNode)
{
	AVLNode<T> *temp;
	temp=parent_AVLNode->right;
	parent_AVLNode->right=left_left_rotation(temp);
	return right_right_rotation(parent_AVLNode);
}

/*...Balancing The AVL-Tree, and Inserting the values/\.
Here we take into account the property of an AVL-Tree, to check whether or not 
the left child has atmost 1 more height than right, if this property is void, it is'nt 
an AVL-Tree, and hence a succession of rotations have to be performed, to make it one,
And then the value is inserted.PS: since the height that can be achieved is only +2 or -2,
we use direct checking properties.
*/

template<class T>
AVLNode<T> *AVLTree<T>::insert(AVLNode<T> *temp, const T &key)  //...Same numbers are Omitted and not inserted.
{
	//...Initial-Condition, Creation of AVL-Tree
	if(temp==NULL)
	{
		temp=new AVLNode<T>();
		temp->data=key;
		temp->left=NULL;
		temp->right=NULL;
		root=temp;//Root points to first AVLNode. Important for Remoe function.	
	}
	else
	{
		/*...The Binary_Search_Tree property, 
			left side smaller, right side greater.*/
		if(key>temp->data)  			
		{
			temp->right=insert(temp->right,key);  //..Recursion.
			if(height_difference(temp)==-2)
			{
				if(key>temp->right->data)
					temp=right_right_rotation(temp);
				else
					temp=right_left_rotation(temp);
			}

		}
		else
		if(key<temp->data)
		{
			temp->left=insert(temp->left,key);  //...Recursion.
			if(height_difference(temp)==2)
			{
				if(key<temp->left->data)
					temp=left_left_rotation(temp);
				else
					temp=left_right_rotation(temp);
			}
		}
	}
	temp->ht=height(temp);
	return(temp);
}
template<class T>
void *AVLTree<T>::inorder(AVLNode<T> *root)  //...Using Inorder Traversal/
{
	//Inorder printing:
	AVLNode<T> *temp=root;
	if(temp!=NULL)
	{
		inorder(temp->left);
		cout<<"\n"<<temp->data<<"[ BF= "<<height_difference(temp)<<" ] ";
		inorder(temp->right);
	}
}
template<class T>
AVLNode<T> *AVLTree<T>::lookup(AVLNode<T> *root, const T &key)  //...Using Inorder Traversal/
{
	T x=key;
	int flag=0;//to check condition in the end.(returning value)
	//Inorder printing:
	AVLNode<T> *temp=root;
	if(temp!=NULL)		
	{
		lookup(temp->left,x);
		if(temp->data==x)
			{
				flag=1;
				cout<<"\nThe Number Was Found in The AVL_Tree";
				return(temp);				
			}
		//cout<<"\n"<<temp->data<<"[ BF= "<<height_difference(temp)<<" ] ";
		lookup(temp->right,x);
	}
	if(!flag)  //...Checks for the 0 condition of the Flag-Variable.
	{
		return(0);
	}
}

template<class T>
AVLNode<T> *AVLTree<T>::remove(AVLNode<T> *temp , const T &key)
{
	AVLNode<T> *p;
	if(temp==NULL)
		{return 0;}
	else
		if(key>temp->data)
			{
				temp->right=remove(temp->right,key);
				if(height_difference(temp)==2)
				{
					if(height_difference(temp->left)>=0)
						temp=left_left_rotation(temp);
					else
						temp=left_right_rotation(temp);
				}

			}
			else
				if(key<temp->data)
				{
					temp->left=remove(temp->left,key);
					if (height_difference(temp)==-2)	
					{
							if(height_difference(temp->right)<=0)
							temp=right_right_rotation(temp);
							else
								temp=right_left_rotation(temp);
					}
				}
				else
				{
					if(temp->right!=NULL)
					{
						p=temp->right;
						while(p->left!=NULL)
							{p=p->left;}
						temp->data=p->data;	
						temp->right=remove(temp->right,p->data);
						if(height_difference(temp)==2)
						{
							if(height_difference(temp->left)>=0)
								temp=left_left_rotation(temp);
								else
								temp=left_right_rotation(temp);

						}
					}
						else
						return(temp->left);

				}
				temp->ht=height(temp);
				return(temp);

}




int main()
{

	AVLTree<int> avl;
	int item;
	int choice;
	char ch='y';
	AVLNode<int> *root=NULL;
	AVLNode<int> *temp=NULL;

	do
	{

		cout<<"\nPlease Enter The umber corresponding to the Choice:";
		cout<<"\n1)Insert Element in the AVL_Tree.\t2)Display the Tree(inorder)\n3)Search for an element.\t4)Delete an Element";
		cin>>choice;

		switch(choice)
		{
			case 1: cout<<"\nPlease Insert The Number to Enter:";
					cin>>item;
					root=avl.insert(root,item);
					break;
			case 2: cout<<"\nThe Inorder Display of the tree is:";
					avl.inorder(root);
					break;
			case 3:cout<<"\n Please Enter the Element to search:";
			       cin>>item;
			       temp=avl.lookup(root,item);//To Print The Node Value.
			       if(temp)
			       {
			       	cout<<"\n"<<temp->data;
			       }

			       break;
			case 4: cout<<"\nEnter the number to delete:";
					cin>>item;
					avl.remove(root,item);
					//avl.balance_tree(root);
					cout<<"\nThe AVL Tree after deletion is:\n";
					avl.inorder(root);
					break;
			default: cout<<"\n Please enter the correct number.";
		}

		cout<<"\nDo You wish to cotinue: (y/n)";
		cin>>ch;
	}while(ch=='Y'||ch=='y');

	/*
	for(int i=0; i<6; i++)
	{
		cout<<"\n Please enter the item:";
		cin>>item;
		root=avl.insert(root,item);
	}
	int a=2;
	avl.inorder(root);



	cout<<"\n\n";
	cout<<"\nPlease Enter the number to show, stuff";
	int x;
	cin>>x;
	avl.remove(a);
	cout<<"\n\n";
	avl.inorder(root);
*/

	return 0;
}

